sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/sns/SnsIO.java

```diff
@@ -257,12 +269,37 @@ public Write withResultOutputTag(TupleTag<PublishResult> results) {
       return builder().setResultOutputTag(results).build();
     }
 
+    /**
+     * Encode the full {@code PublishResult} object, including sdkResponseMetadata and
+     * sdkHttpMetadata but excluding the HTTP response headers.
+     */
+    public Write withFullPublishResultNoHeaders() {
+      return builder().setCoder(PublishResultCoders.fullPublishResultNoHeaders()).build();
+    }
+
+    /**
+     * Encode the full {@code PublishResult} object, including sdkResponseMetadata and
+     * sdkHttpMetadata with the HTTP response headers.
+     */
+    public Write withFullPublishResultIncludingHeaders() {
+      return builder().setCoder(PublishResultCoders.fullPublishResultIncludingHeaders()).build();
+    }
+
+    /** Encode the {@code PublishResult} with the given coder. */
+    public Write withPublishResultCoder(Coder<PublishResult> coder) {
+      return builder().setCoder(coder).build();
+    }
+
     @Override
     public PCollectionTuple expand(PCollection<PublishRequest> input) {
       checkArgument(getTopicName() != null, "withTopicName() is required");
-      return input.apply(
-          ParDo.of(new SnsWriterFn(this))
-              .withOutputTags(getResultOutputTag(), TupleTagList.empty()));
+      PCollectionTuple result =
+          input.apply(
+              ParDo.of(new SnsWriterFn(this))
+                  .withOutputTags(getResultOutputTag(), TupleTagList.empty()));
+      Optional.ofNullable(getCoder())
+          .ifPresent(coder -> result.get(getResultOutputTag()).setCoder(coder));
```

Comment:
nit: It is much easier for someone to understand the following logic:
```suggestion
      if (getCoder() != null) {
        result.get(getResultOutputTag()).setCoder(coder);
      }
```